File: Problem2.py
from collections import deque
class SnakeGame:

    def __init__(self, width: int, height: int, food: List[List[int]]):
        self.snake=deque()
        self.headptr=[0,0]
        self.snake.append((0,0))
        self.foodList=food
        self.h=height
        self.w=width
        self.i=0

    def move(self, direction: str) -> int:
        if direction=='U':
            self.headptr[0]-=1
        elif direction=='D':
            self.headptr[0]+=1
        elif direction=='R':
            self.headptr[1]+=1
        else:
            self.headptr[1]-=1

        if self.headptr[0]<0 or self.headptr[0]>=self.h or self.headptr[1]<0 or self.headptr[1]>=self.w:
            return -1

        tr = self.snake[-1][0]
        tc = self.snake[-1][1]

        if (self.headptr[0],self.headptr[1]) in self.snake and (self.headptr[0],self.headptr[1])!=(tr,tc):
            return -1
        
        if self.i<len(self.foodList):
            currFood=self.foodList[self.i]
            if currFood[0]==self.headptr[0] and currFood[1]==self.headptr[1]:
                self.snake.appendleft((self.headptr[0],self.headptr[1]))
                self.i+=1
                return len(self.snake)-1

        tail=self.snake.pop()
        self.snake.appendleft((self.headptr[0],self.headptr[1]))

        return len(self.snake)-1


        


# Your SnakeGame object will be instantiated and called as such:
# obj = SnakeGame(width, height, food)
# param_1 = obj.move(direction)

File: problem2.java
//TC:O(m*n)
//SC:O(m*n)

class SnakeGame {
    int width;
    int height;
    LinkedList<int[]> snakebody;
    LinkedList<int[]> fooditems;
    int[] snakehead;

    public SnakeGame(int width, int height, int[][] food) {
        this.width = width;
        this.height = height;
        snakebody = new LinkedList<>();
        fooditems = new LinkedList<>(Arrays.asList(food));
        snakehead = new int[]{0, 0};
        snakebody.addLast(snakehead);
    }
    
    public int move(String direction) {
        if(direction.equals("U")){
            snakehead[0]--;
        }else if(direction.equals("D")){
            snakehead[0]++;
        }else if(direction.equals("L")){
            snakehead[1]--;
        }else if(direction.equals("R")){
            snakehead[1]++;
        }
        
        if(snakehead[0] <0 || snakehead[0] == height || snakehead[1] < 0 || snakehead[1] == width){
            return -1;
        }
        
        for(int i = 1; i<snakebody.size(); i++){
            int[] curr = snakebody.get(i);
            if(curr[0] == snakehead[0] && curr[1] == snakehead[1]) {
                return -1;
            }
        }
        
        if(fooditems.size() > 0){
           // for(int i = 0; i < fooditems.size(); i++){
                int[] fooditem = fooditems.get(0);
                if(fooditem[0] == snakehead[0] && fooditem[1] == snakehead[1]){
                    snakebody.addLast(new int[] {snakehead[0], snakehead[1]});
                    fooditems.removeFirst();
                    return snakebody.size() - 1;
                }
           // }
        }
        snakebody.addLast(new int[] {snakehead[0], snakehead[1]});
        snakebody.removeFirst();
        return snakebody.size() -1;
    }
}

/**
 * Your SnakeGame object will be instantiated and called as such:
 * SnakeGame obj = new SnakeGame(width, height, food);
 * int param_1 = obj.move(direction);
 */
